sklearn/metrics/pairwise.py

```diff
@@ -687,7 +613,245 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def gower_distances(X, Y=None, categorical_features=None, scale=True, **metric_kwargs):
+    """Compute the distances between the observations in X and Y,
+    that may contain mixed types of data, using an implementation
+    of Gower formula.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    Y : array-like, or pandas.DataFrame, optional,
+        shape (n_samples, n_features)
+
+    categorical_features : array-like, optional, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+        This is useful when categorical atributes are represented as integer
+        values. Categorical ordinal attributes are treated as numeric, and
+        must be marked as false.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+        If the categorical_features array is not provided, by default all
+        non-numeric columns are considered categorical.
+
+    scale : boolean, optional (default=True)
```

Comment:
I think the way to do `_precompute_metric_params` is to allow `scale` to be _either_ a bool or an array-like of shape `(n_features,)` like the `scale_` attribute of MinMaxScaler. Then:
* In `gower_distances`:
    * if scale is an array: the distances between numeric variables of `x` and `y` are like `sum((x - y) / scale)`
    * if scale is True: it should be equivalent to `scale=np.ptp(X, axis=0)`
    * if scale is False: it should be equivalent to `scale=np.ones(n_features)`

* In `_precompute_metric_params`:
    * if scale is an array: do not change it
    * if scale is True: it should be equivalent to `scale=np.ptp(X, axis=0)`
    * if scale is unset: it should be equivalent to `scale=np.ptp(X, axis=0)` (since scale=True is default)
    * if scale is False: do not change it

I've not detailed how to handle categorical features, but I hope it is reasonably straightforward to define their handling.

Reply:
In the other methods (seuclidean, mahalanobis) _precompute_metric_params uses a dictionary (kwargs), should we use it for a list of ranges? I suggest to bring this code back to gower. 

>> I've not detailed how to handle categorical features, but I hope it is reasonably
>> straightforward to define their handling.

We don't need to scale categorical features, only the numerical ones. I need to make it clear in the documentation.

Reply:
I'll pass the array of ranges to kwargs in the _precompute_metric_params

Reply:
I'm not sure what you're asking. The job of `_precompute_metric_params` is to take a dict of args and return a dict of changed/new args:
https://github.com/scikit-learn/scikit-learn/blob/f23e92ed4cdc5a952331e597023bd2c9922e6f9d/sklearn/metrics/pairwise.py#L1431-L1432

Reply:
I'm not sure what you're asking. The job of `_precompute_metric_params` is to take a dict of args and return a dict of changed/new args:
https://github.com/scikit-learn/scikit-learn/blob/f23e92ed4cdc5a952331e597023bd2c9922e6f9d/sklearn/metrics/pairwise.py#L1431-L1432